File: frontend/src/api/benchmark.ts
import { get, post } from "./client";

export interface BenchmarkTaskSummary {
  id: string;
  source: string;
  category: string;
  answer_type: string;
  word_count: number | null;
  question_preview: string;
}

export interface BenchmarkCondition {
  code: string;
  name: string;
  deprecated?: boolean;
}

export interface BenchmarkRunSummary {
  run_id: string;
  config: {
    task_ids: string[] | null;
    models: string[];
    conditions: string[];
    runs_per_task: number;
    skip_completed: boolean;
    prompt_version: string;
    execution_seed: number;
    stage: string;
  };
  status: string;
  created_at: string;
  completed_at: string | null;
  total_items: number;
  completed_items: number;
  correct_items: number;
  total_cost_usd: number;
  total_tokens: number;
  accuracy_by_condition: Record<string, number>;
  dropouts: Array<{ task_id: string; model: string; condition: string; error: string }>;
}

export interface BenchmarkResult {
  task_id: string;
  condition: string;
  model_id: string;
  run_number: number;
  timestamp: string;
  latency_seconds: number;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  raw_response: string;
  extracted_answer: string | null;
  is_correct: boolean | null;
  correct_answer: string;
  answer_type: string;
  question: string | null;
  source: string | null;
}

export interface AccuracySummary {
  total: number;
  total_correct: number;
  overall_accuracy: number;
  by_condition: Record<string, { correct: number; total: number; accuracy: number; name: string }>;
  by_source: Record<string, { correct: number; total: number; accuracy: number }>;
  by_model: Record<string, { correct: number; total: number; accuracy: number }>;
}

export function fetchBenchmarkTasks(taskVersion?: string): Promise<BenchmarkTaskSummary[]> {
  const qs = taskVersion ? `?task_version=${taskVersion}` : "";
  return get<BenchmarkTaskSummary[]>(`/benchmark/tasks${qs}`);
}

export function fetchBenchmarkConditions(includeDeprecated: boolean = true): Promise<BenchmarkCondition[]> {
  return get<BenchmarkCondition[]>(`/benchmark/conditions?include_deprecated=${includeDeprecated}`);
}

export function fetchBenchmarkRuns(): Promise<BenchmarkRunSummary[]> {
  return get<BenchmarkRunSummary[]>("/benchmark/runs");
}

export function fetchBenchmarkRun(runId: string): Promise<BenchmarkRunSummary> {
  return get<BenchmarkRunSummary>(`/benchmark/runs/${runId}`);
}

export function fetchBenchmarkRunResults(runId: string): Promise<BenchmarkResult[]> {
  return get<BenchmarkResult[]>(`/benchmark/runs/${runId}/results`);
}

export function startBenchmarkRun(config: {
  task_ids?: string[] | null;
  models: string[];
  conditions: string[];
  runs_per_task?: number;
  skip_completed?: boolean;
  stage?: string;
  task_version?: string;
}): Promise<{ run_id: string; total_items: number; models: string[]; conditions: string[]; stage: string }> {
  return post("/benchmark/runs", config);
}

export function cancelBenchmarkRun(runId: string): Promise<{ status: string }> {
  return post(`/benchmark/runs/${runId}/cancel`, {});
}

export function fetchAccuracySummary(params?: { model_id?: string; prompt_version?: string }): Promise<AccuracySummary> {
  const qs = new URLSearchParams();
  if (params?.model_id) qs.set("model_id", params.model_id);
  if (params?.prompt_version) qs.set("prompt_version", params.prompt_version);
  const suffix = qs.toString() ? `?${qs}` : "";
  return get<AccuracySummary>(`/benchmark/accuracy${suffix}`);
}

export interface BenchmarkEstimate {
  total_calls: number;
  models: Array<{ model: string; calls: number; estimated_cost_usd: number }>;
  benchmark_cost_usd: number;
  error_classification: {
    estimated_incorrect: number;
    classification_calls: number;
    scorers: string[];
    estimated_cost_usd: number;
  };
  total_estimated_cost_usd: number;
}

export interface ErrorClassificationResult {
  total_results: number;
  incorrect: number;
  classifications: number;
  file: string;
  agreement: {
    kappa: number | null;
    exact_agreement: number;
    n_pairs: number;
    exact_matches: number;
    raters: string[];
  };
  by_rater: Record<string, Record<string, number>>;
  by_condition: Record<string, Record<string, number>>;
}

export function classifyErrors(config: {
  run_id?: string;
  scorer_a?: string;
  scorer_b?: string;
}): Promise<ErrorClassificationResult> {
  return post("/benchmark/classify-errors", config);
}

export function fetchErrorClassifications(): Promise<Array<{ file: string; count: number; timestamp: string }>> {
  return get("/benchmark/error-classifications");
}

export function fetchErrorClassificationDetail(filename: string): Promise<{
  classifications: Array<{
    task_id: string;
    condition: string;
    model_id: string;
    classification: string;
    reasoning: string;
    root_cause_quote: string | null;
    confidence: string;
    rater: string;
  }>;
  agreement: { kappa: number | null; exact_agreement: number; n_pairs: number; raters: string[] };
  by_condition: Record<string, Record<string, number>>;
  total: number;
}> {
  return get(`/benchmark/error-classifications/${filename}`);
}

export function fetchAnalysis(params?: { run_id?: string }): Promise<Record<string, any>> {
  const qs = new URLSearchParams();
  if (params?.run_id) qs.set("run_id", params.run_id);
  const suffix = qs.toString() ? `?${qs}` : "";
  return get(`/benchmark/analysis${suffix}`);
}

export function estimateBenchmarkCost(config: {
  task_ids?: string[] | null;
  models: string[];
  conditions: string[];
  runs_per_task?: number;
}): Promise<BenchmarkEstimate> {
  return post("/benchmark/estimate", config);
}

File: backend/dodar/models/benchmark.py
"""Data models for Phase 2 benchmark tasks and results."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel


class AnswerType(str, Enum):
    MULTIPLE_CHOICE = "multiple_choice"
    NUMERIC_EXACT = "numeric_exact"
    NUMERIC = "numeric"  # alias used in v2 task bank
    EXACT_MATCH = "exact_match"


class BenchmarkTask(BaseModel):
    """A single benchmark task. Supports v1, v2, and v3 formats."""

    id: str
    source: str
    question: str
    correct_answer: str
    answer_type: AnswerType

    # Options — three formats:
    #   v1: dict[str, str] e.g. {"A": "...", "B": "..."}
    #   v2: None (options inline in question text)
    #   v3: list[str] with separate option_labels
    options: dict[str, str] | list[str] | None = None
    option_labels: list[str] | None = None

    # v1 fields
    source_index: int | None = None
    category: str | None = None
    correct_answer_text: str | None = None
    reasoning_steps: int | None = None
    word_count: int | None = None

    # v3 fields
    scoring: dict | None = None
    provenance: dict | None = None
    selection_notes: list[str] | None = None

    @property
    def effective_answer_type(self) -> str:
        """Normalise answer type — 'numeric' treated as 'numeric_exact'."""
        if self.answer_type == AnswerType.NUMERIC:
            return "numeric_exact"
        return self.answer_type.value

    @property
    def formatted_options(self) -> str | None:
        """Format options for prompt insertion, regardless of storage format."""
        if self.options is None:
            return None
        if isinstance(self.options, dict):
            # v1: {"A": "text", "B": "text"}
            return "\n".join(f"  {k}: {v}" for k, v in self.options.items())
        if isinstance(self.options, list):
            # v3: ["text1", "text2", ...] with option_labels
            labels = self.option_labels or [chr(65 + i) for i in range(len(self.options))]
            return "\n".join(f"  {l}: {o}" for l, o in zip(labels, self.options))
        return None


class BenchmarkTaskSet(BaseModel):
    """Top-level wrapper for the benchmark tasks JSON file."""

    metadata: dict
    tasks: list[BenchmarkTask]


# --- Experimental conditions (Phase 2 protocol) ---

class ConditionCode(str, Enum):
    BASELINE = "A"
    ZERO_SHOT_COT = "B"
    PGR = "C"
    REACT = "D"
    STEP_BACK = "E"
    SHUFFLED_PGR = "F"
    FEW_SHOT_COT = "G"
    ANTI_ANCHORING_PGR = "H"


CONDITION_NAMES: dict[str, str] = {
    "A": "Baseline",
    "B": "Zero-Shot CoT",
    "C": "PGR (Late Commit)",
    "C_previous": "PGR v2 (Early Commit)",
    "D": "ReAct",
    "E": "Step-Back",
    "F": "Shuffled PGR",
    "G": "Few-Shot CoT",
    "H": "Anti-Anchoring PGR",
}


# --- Benchmark run results ---

class BenchmarkResult(BaseModel):
    """Result of running one benchmark task under one condition with one model."""

    task_id: str
    condition: str  # condition code: A-G
    model_id: str
    run_number: int = 1  # for multi-run stages (1, 2, or 3)
    prompt_version: str = "v3.2"

    # Timing and cost
    timestamp: datetime
    latency_seconds: float
    input_tokens: int
    output_tokens: int
    cost_usd: float

    # Prompt sent (for debugging/audit)
    system_prompt_sent: str | None = None
    user_prompt_sent: str

    # Response
    raw_response: str
    extracted_answer: str | None = None
    is_correct: bool | None = None

    # Ground truth (copied from task for convenience)
    correct_answer: str
    answer_type: str

    # Metadata
    question: str | None = None  # stored for blinding script convenience
    source: str | None = None


class BenchmarkRunConfig(BaseModel):
    """Configuration for a benchmark run."""

    task_ids: list[str] | None = None  # None = all 100 tasks
    models: list[str]
    conditions: list[str]  # condition codes: ["A", "B", "C", ...]
    runs_per_task: int = 1
    skip_completed: bool = True
    prompt_version: str = "v3.2"
    execution_seed: int = 42  # for shuffling execution order
    stage: Literal["triage", "validate", "full"] = "triage"
    task_version: str | None = None  # "v1" or "v2" (None = default)


class BenchmarkRunSummary(BaseModel):
    """Summary of a benchmark run."""

    run_id: str
    config: BenchmarkRunConfig
    status: str = "pending"
    created_at: datetime
    completed_at: datetime | None = None

    total_items: int = 0
    completed_items: int = 0
    correct_items: int = 0
    total_cost_usd: float = 0.0
    total_tokens: int = 0

    # Accuracy by condition
    accuracy_by_condition: dict[str, float] = {}
    # Dropout tracking
    dropouts: list[dict] = []


# --- Error classification ---

class ErrorCategory(str, Enum):
    PREMATURE_CLOSURE = "PREMATURE_CLOSURE"
    ANCHORING_ERROR = "ANCHORING_ERROR"
    INCOMPLETE_SEARCH = "INCOMPLETE_SEARCH"
    FAILURE_TO_REVISE = "FAILURE_TO_REVISE"
    EXECUTION_ERROR = "EXECUTION_ERROR"
    COMPREHENSION_FAILURE = "COMPREHENSION_FAILURE"
    ABSTENTION = "ABSTENTION"


class ErrorClassification(BaseModel):
    """Error classification for an incorrect response."""

    task_id: str
    condition: str
    model_id: str
    run_number: int = 1

    classification: ErrorCategory
    reasoning: str
    root_cause_quote: str | None = None
    confidence: Literal["high", "medium", "low"] = "medium"
    rater: str = "human"  # "human", "claude-opus-4-6", "gpt-5.4"
